driver-core/src/main/com/mongodb/internal/connection/IamAuthenticator.java

```diff
@@ -0,0 +1,275 @@
+                    .setTimestamp(timestamp)
+                    .build();
+
+            BsonDocument ret = new BsonDocument()
+                    .append("a", new BsonString(authorizationHeader.toString()))
+                    .append("d", new BsonString(authorizationHeader.getTimestamp()));
+            if (token != null) {
+                ret.append("t", new BsonString(token));
+            }
+
+            return toBson(ret);
+        }
+
+
+        private byte[] toBson(final BsonDocument document) {
+            byte[] bytes;
+            BasicOutputBuffer buffer = new BasicOutputBuffer();
+            new BsonDocumentCodec().encode(new BsonBinaryWriter(buffer), document, EncoderContext.builder().build());
+            bytes = new byte[buffer.size()];
+            int curPos = 0;
+            for (ByteBuf cur : buffer.getByteBuffers()) {
+                System.arraycopy(cur.array(), cur.position(), bytes, curPos, cur.limit());
+                curPos += cur.position();
+            }
+            return bytes;
+        }
+
+        @NonNull
+        String getUserName() {
+            String userName = credential.getUserName();
+            if (userName == null) {
+                userName = BsonDocument
+                        .parse(getHttpResponse())
+                        .getString("AccessKeyId")
+                        .getValue();
+            }
+            return userName;
+        }
+
+        @NonNull
+        private String getPassword() {
+            char[] password = credential.getPassword();
+            if (password == null) {
+                password = BsonDocument
+                        .parse(getHttpResponse())
+                        .getString("SecretAccessKey")
+                        .getValue()
+                        .toCharArray();
+            }
+            return new String(password);
+        }
+
+        @Nullable
+        private String getSessionToken() {
+            String token = credential.getMechanismProperty("AWS_SESSION_TOKEN", null);
+            if (credential.getPassword() == null || credential.getUserName() == null) {
+                if (token == null) {
+                    token = BsonDocument
+                            .parse(getHttpResponse())
+                            .getString("Token")
+                            .getValue();
+                } else {
+                    throw new IllegalArgumentException("The connection string contains auth properties and no username and password");
+                }
+            }
+            return token;
+        }
+
+
+        @NonNull
+        private String getHttpResponse() {
+            if (httpResponse == null) {
+                final String ec2 = "http://169.254.169.254/latest/meta-data/iam/security-credentials/";
+                final String ecs = "http://169.254.170.2";
+
+                String path = System.getenv("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI");
+                String uri = ecs + path;
```

Comment:
Clearer if you only set uri once, like this:

    String uri = (path == null) 
         ? ec2 + getHttpContents(ec2) 
         : ecs + path;

Reply:
Done